src/Orleans/Core/InterceptedMethodInvokerCache.cs

```diff
@@ -34,74 +32,142 @@ internal class InterceptedMethodInvokerCache
+            foreach (var iface in interfaces)
             {
-                IGrainExtension extension;
-                if (extensionMap.TryGetExtension(interfaceId, out extension))
+                var methods = GrainInterfaceUtils.GetMethods(iface);
+
+                // Map every method on this interface from the definition interface onto the implementation class.
+                var methodMap = new Dictionary<int, MethodInfo>(methods.Length);
+                var mapping = default(InterfaceMapping);
+                foreach (var method in methods)
                 {
-                    implementationType = extension.GetType();
+                    // If this method is not from the expected interface (eg, because it's from a parent interface), then
+                    // get the mapping for the interface which it does belong to.
+                    if (mapping.InterfaceType != method.DeclaringType)
+                    {
+                        mapping = implementationTypeInfo.GetRuntimeInterfaceMap(method.DeclaringType);
+                    }
+
+                    // Find the index of the interface method and then get the implementation method at that position.
+                    for (var k = 0; k < mapping.InterfaceMethods.Length; k++)
+                    {
+                        if (mapping.InterfaceMethods[k] != method) continue;
+                        methodMap[GrainInterfaceUtils.ComputeMethodId(method)] = mapping.TargetMethods[k];
+                        break;
+                    }
                 }
+
+                // Add the resulting map of methodId -> method to the interface map.
+                var interfaceId = GrainInterfaceUtils.GetGrainInterfaceId(iface);
+                result[interfaceId] = new InterceptedMethodInvoker(invoker, methodMap);
             }
 
-            var implementationMap = GetInterfaceToImplementationMap(interfaceId, implementationType);
-            return new InterceptedMethodInvoker(invoker, implementationMap);
+            return result;
         }
 
         /// <summary>
-        /// Maps the provided <paramref name="interfaceId"/> to the provided <paramref name="implementationType"/>.
+        /// Maps the interfaces of the provided <paramref name="implementationType"/>.
         /// </summary>
-        /// <param name="interfaceId">The interface id.</param>
-        /// <param name="implementationType">The implementation type.</param>
+        /// <param name="implementationType">The implementation type, which must be a concrete generic type.</param>
+        /// <param name="invoker">The grain method invoker.</param>
         /// <returns>The mapped interface.</returns>
-        private static Dictionary<int, MethodInfo> GetInterfaceToImplementationMap(
-            int interfaceId,
-            Type implementationType)
+        private static Dictionary<int, InterceptedMethodInvoker> GetGenericInterfaceInvoker(
```

Comment:
CreateInterfaceToIntercepterInvokerMap?
from comment "This method creates a map interfaceId -> intercepted invoker"